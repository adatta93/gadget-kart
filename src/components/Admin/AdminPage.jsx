import React, { useEffect, useState } from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import Modal from "../../ConfirmModal";
import Loading from "../../Loading";
import {
  deleteProduct,
  fetchAllProducts,
} from "../../redux/products/products.action";

export function AdminPage({ products, fetchAllProducts, deleteProduct }) {
  const [selectedProdForDelete, setSelectedProdForDelete] = useState(null);
  const [modal, setModal] = useState(false);

  useEffect(() => {
    fetchAllProducts();
  }, [fetchAllProducts]);

  const handleDeleteClick = (id) => {
    setSelectedProdForDelete(id);
    setModal(true);
  };

  const handleSubmit = (res) => {
    console.log("modal resp ", res);
    if (res === "yes") {
      deleteProduct(selectedProdForDelete);
    }
    modalClose();
  };

  const modalClose = () => {
    setModal(false);
  };

  return (
    <div className="container my-4">
      <div className="row align-items-center">
        <div className="col-md-6">
          <h2>Product Administration</h2>
        </div>
        <div className="col-md-6 d-flex justify-content-md-end">
          <Link className="btn btn-login" to="/admin/add-product">
            Add New Product
          </Link>
        </div>
      </div>
      <table className="table table-striped admin-table">
        <thead>
          <tr>
            <th>Product Name</th>
            <th>Category</th>
            <th className="text-right">Price</th>
            <th></th>
          </tr>
        </thead>
        <tbody>
          {products ? (
            products.map((prod) => (
              <tr key={prod.id}>
                <td>{prod.title}</td>
                <td>{prod.category}</td>
                <td className="text-right">
                  {prod.price.toLocaleString(undefined, {
                    minimumFractionDigits: 2,
                    style: "currency",
                    currency: "INR",
                  })}
                </td>
                <td className="text-center">
                  <Link
                    className="btn btn-info prod-edit"
                    to={{
                      pathname: "/admin/edit-product",
                      state: { product: prod },
                    }}>
                    <i className="fas fa-edit"></i>
                    <span className="ml-2 d-none d-md-inline-block">Edit</span>
                  </Link>
                  <button
                    className="btn btn-danger ml-md-2"
                    onClick={() => handleDeleteClick(prod.id)}>
                    <i className="fas fa-trash"></i>
                    <span className="ml-2 d-none d-md-inline-block">
                      Delete
                    </span>
                  </button>
                </td>
              </tr>
            ))
          ) : (
            <Loading />
          )}
        </tbody>
      </table>
      <Modal show={modal}>
        <div className="modal-footer">
          <button
            className="btn btn-secondary"
            onClick={() => handleSubmit("no")}>
            Cancel
          </button>
          <button
            className="btn btn-danger"
            onClick={() => handleSubmit("yes")}>
            DELETE
          </button>
        </div>
      </Modal>
    </div>
  );
}

const mapStateToProps = (state) => {
  console.log("DTL ", state);
  return {
    products: state.products.allProducts,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    fetchAllProducts: () => dispatch(fetchAllProducts),
    deleteProduct: (id) => dispatch(deleteProduct(id)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(AdminPage);
