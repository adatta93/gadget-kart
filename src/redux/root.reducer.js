import { combineReducers } from "redux";
import { cartReducer } from "./cart/cart.reducer";
import { fetchProductReducer } from "./products/products.reducer";
import { userReducer } from "./users/users.reducer";

export default combineReducers({
  products: fetchProductReducer,
  user: userReducer,
  cart: cartReducer,
});
