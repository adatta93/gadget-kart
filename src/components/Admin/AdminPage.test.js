import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import Enzyme, { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { createBrowserHistory } from "history";
import React from "react";
import { Provider } from "react-redux";
import { Router } from "react-router-dom";
import { applyMiddleware, createStore } from "redux";
import thunk from "redux-thunk";
import rootReducer from "../../redux/root.reducer";
import { AdminPage } from "./AdminPage";

Enzyme.configure({ adapter: new Adapter() });

let store;
let props;
let history;
beforeAll(() => {
  props = {
    products: [
      {
        id: 15,
        title: "Lenovo Ideapad S145 Core i3 7th Gen",
        img: "lenovo-ideapad.jpeg",
        category: "laptop",
        price: 31000,
        processor: "Intel Core i3 7th Gen",
        ram: "4GB",
        memory: "1TB HDD",
        screen: "15.6in Display",
        weight: "1.85kg",
      },
    ],
    fetchAllProducts: jest.fn(),
    deleteProduct: jest.fn(),
  };
  store = createStore(rootReducer, applyMiddleware(thunk));
  history = createBrowserHistory();
});

describe("AdminPage Snapshot", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <AdminPage {...props} />
        </Router>
      </Provider>
    );
  });

  it("renders correctly", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  afterEach(() => {
    wrapper.unmount();
  });
});

describe("When valid products array props passed to AdminPage component", () => {
  let wrapper;
  let props;
  beforeAll(() => {
    props = {
      products: [
        {
          id: 15,
          title: "Lenovo Ideapad S145 Core i3 7th Gen",
          img: "lenovo-ideapad.jpeg",
          category: "laptop",
          price: 31000,
          processor: "Intel Core i3 7th Gen",
          ram: "4GB",
          memory: "1TB HDD",
          screen: "15.6in Display",
          weight: "1.85kg",
        },
      ],
    };
    wrapper = shallow(<AdminPage {...props} />);
  });

  it('renders "Product Name" as heading in first column', () => {
    expect(wrapper.find("th").at(0).render().text()).toEqual("Product Name");
  });

  it("renders correct Price value in third column of first row", () => {
    expect(wrapper.find("tr").at(1).find("td").at(2).render().text()).toEqual(
      "â‚¹31,000.00"
    );
  });
});
