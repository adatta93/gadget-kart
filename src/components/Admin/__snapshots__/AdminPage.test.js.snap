// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminPage Snapshot renders correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <AdminPage
      deleteProduct={[MockFunction]}
      fetchAllProducts={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      products={
        Array [
          Object {
            "category": "laptop",
            "id": 15,
            "img": "lenovo-ideapad.jpeg",
            "memory": "1TB HDD",
            "price": 31000,
            "processor": "Intel Core i3 7th Gen",
            "ram": "4GB",
            "screen": "15.6in Display",
            "title": "Lenovo Ideapad S145 Core i3 7th Gen",
            "weight": "1.85kg",
          },
        ]
      }
    >
      <div
        className="container my-4"
      >
        <div
          className="row align-items-center"
        >
          <div
            className="col-md-6"
          >
            <h2>
              Product Administration
            </h2>
          </div>
          <div
            className="col-md-6 d-flex justify-content-md-end"
          >
            <Link
              className="btn btn-login"
              to="/admin/add-product"
            >
              <LinkAnchor
                className="btn btn-login"
                href="/admin/add-product"
                navigate={[Function]}
              >
                <a
                  className="btn btn-login"
                  href="/admin/add-product"
                  onClick={[Function]}
                >
                  Add New Product
                </a>
              </LinkAnchor>
            </Link>
          </div>
        </div>
        <table
          className="table table-striped admin-table"
        >
          <thead>
            <tr>
              <th>
                Product Name
              </th>
              <th>
                Category
              </th>
              <th
                className="text-right"
              >
                Price
              </th>
              <th />
            </tr>
          </thead>
          <tbody>
            <tr
              key="15"
            >
              <td>
                Lenovo Ideapad S145 Core i3 7th Gen
              </td>
              <td>
                laptop
              </td>
              <td
                className="text-right"
              >
                â‚¹31,000.00
              </td>
              <td
                className="text-center"
              >
                <Link
                  className="btn btn-info prod-edit"
                  to={
                    Object {
                      "pathname": "/admin/edit-product",
                      "state": Object {
                        "product": Object {
                          "category": "laptop",
                          "id": 15,
                          "img": "lenovo-ideapad.jpeg",
                          "memory": "1TB HDD",
                          "price": 31000,
                          "processor": "Intel Core i3 7th Gen",
                          "ram": "4GB",
                          "screen": "15.6in Display",
                          "title": "Lenovo Ideapad S145 Core i3 7th Gen",
                          "weight": "1.85kg",
                        },
                      },
                    }
                  }
                >
                  <LinkAnchor
                    className="btn btn-info prod-edit"
                    href="/admin/edit-product"
                    navigate={[Function]}
                  >
                    <a
                      className="btn btn-info prod-edit"
                      href="/admin/edit-product"
                      onClick={[Function]}
                    >
                      <i
                        className="fas fa-edit"
                      />
                      <span
                        className="ml-2 d-none d-md-inline-block"
                      >
                        Edit
                      </span>
                    </a>
                  </LinkAnchor>
                </Link>
                <button
                  className="btn btn-danger ml-md-2"
                  onClick={[Function]}
                >
                  <i
                    className="fas fa-trash"
                  />
                  <span
                    className="ml-2 d-none d-md-inline-block"
                  >
                    Delete
                  </span>
                </button>
              </td>
            </tr>
          </tbody>
        </table>
        <Modal
          show={false}
        >
          <div
            className="modal d-none"
          >
            <div
              className="modal-dialog modal-dialog-centered"
            >
              <div
                className="modal-content"
              >
                <div
                  className="modal-header"
                >
                  <h5
                    className="modal-title"
                  >
                    Confirm
                  </h5>
                </div>
                <div
                  className="modal-body"
                >
                  <p>
                    Are you sure you want to delete this product?
                  </p>
                </div>
                <div
                  className="modal-footer"
                >
                  <button
                    className="btn btn-secondary"
                    onClick={[Function]}
                  >
                    Cancel
                  </button>
                  <button
                    className="btn btn-danger"
                    onClick={[Function]}
                  >
                    DELETE
                  </button>
                </div>
              </div>
            </div>
          </div>
        </Modal>
      </div>
    </AdminPage>
  </Router>
</Provider>
`;
